server/src/main/java/org/elasticsearch/common/network/CloseableChannel.java

```diff
@@ -0,0 +1,115 @@
+
+    /**
+     * Closes the channel. This might be an asynchronous process. There is no guarantee that the channel
+     * will be closed when this method returns. Use the {@link #addCloseListener(ActionListener)} method
+     * to implement logic that depends on knowing when the channel is closed.
+     */
+    @Override
+    void close();
+
+    /**
+     * Adds a listener that will be executed when the channel is closed. If the channel is still open when
+     * this listener is added, the listener will be executed by the thread that eventually closes the
+     * channel. If the channel is already closed when the listener is added the listener will immediately be
+     * executed by the thread that is attempting to add the listener.
+     *
+     * @param listener to be executed
+     */
+    void addCloseListener(ActionListener<Void> listener);
+
+    /**
+     * Indicates whether a channel is currently open
+     *
+     * @return boolean indicating if channel is open
+     */
+    boolean isOpen();
+
+    /**
+     * Closes the channel without blocking.
+     *
+     * @param channel to close
+     */
+    static <C extends CloseableChannel> void closeChannel(C channel) {
+        closeChannel(channel, false);
+    }
+
+    /**
+     * Closes the channel.
+     *
+     * @param channel to close
+     * @param blocking indicates if we should block on channel close
+     */
+    static <C extends CloseableChannel> void closeChannel(C channel, boolean blocking) {
+        closeChannels(Collections.singletonList(channel), blocking);
+    }
+
+    /**
+     * Closes the channels.
+     *
+     * @param channels to close
+     * @param blocking indicates if we should block on channel close
+     */
+    static <C extends CloseableChannel> void closeChannels(List<C> channels, boolean blocking) {
+        if (blocking) {
+            ArrayList<ActionFuture<Void>> futures = new ArrayList<>(channels.size());
+            for (final C channel : channels) {
+                if (channel.isOpen()) {
+                    PlainActionFuture<Void> closeFuture = PlainActionFuture.newFuture();
+                    channel.addCloseListener(closeFuture);
+                    channel.close();
+                    futures.add(closeFuture);
+                }
+            }
+            blockOnFutures(futures);
+        } else {
+            IOUtils.closeWhileHandlingException(channels);
+        }
+    }
+
+    static void blockOnFutures(List<ActionFuture<Void>> futures) {
+        for (ActionFuture<Void> future : futures) {
+            try {
+                future.get();
+            } catch (ExecutionException e) {
+                // Ignore as we are only interested in waiting for the close process to complete. Logging
+                // close exceptions happens elsewhere.
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new IllegalStateException("Future got interrupted", e);
```

Comment:
wouldn't it be ok to just continue if we get interrupted?

Reply:
So just this:

```
    static void blockOnFutures(List<ActionFuture<Void>> futures) {
        for (ActionFuture<Void> future : futures) {
            try {
                future.get();
            } catch (ExecutionException e) {
                // Ignore as we are only interested in waiting for the close process to complete. Logging
                // close exceptions happens elsewhere.
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
```

Reply:
yeah I wonder if that is hte right thing?